deptrack-devops/src/Devops/Graph.hs

```diff
@@ -153,44 +154,48 @@ noBroadcast :: Broadcast
 noBroadcast = const (return ())
 
 -- | Turn-up a graph sequentially
--- Topologically sort all nodes of the graph then run corresponding operations
+--
+-- Topologically sorts a transposition of the graph then invokes `TurnUp` action synchronously
+-- through each node. Note the order of execution is deterministic.
 syncTurnupGraph :: Broadcast -> OpGraph -> IO ()
 syncTurnupGraph bcast (graph,lookupVertex,_) =
-  mapM_ go (reverse $ Graph.topSort graph)
-  where
-    go :: Graph.Vertex -> IO ()
-    go vertex = do
-        let (preop,oid,_) = lookupVertex vertex
-            desc          = opName . opDescription $ runPreOp preop
-            funs          = opFunctions $ runPreOp preop
-            turnup        = opTurnup funs
-            reload        = opReload funs
-            check         = opCheck  funs
-        bcast (oid,Unknown,Transient,TurnedUp)
-        print ("pre-checking: " <> desc)
-        currentStatus <- check
-        case currentStatus of
-            Success -> print ("reloading: " <> desc) >> reload
-            _       -> print ("turning-up: " <> desc) >> turnup
-        print ("turnup-done: " <> desc)
-        bcast (oid,Success,Stable,TurnedUp)
+    mapM_ go (Graph.topSort $ Graph.transposeG graph)
+    where
+      go :: Graph.Vertex -> IO ()
+      go vertex = do
+          let (preop,oid,_) = lookupVertex vertex
+              desc          = opName . opDescription $ runPreOp preop
+              funs          = opFunctions $ runPreOp preop
+              turnup        = opTurnup funs
+              reload        = opReload funs
+              check         = opCheck  funs
+          bcast (oid,Unknown,Transient,TurnedUp)
+          print ("pre-checking: " <> desc)
+          currentStatus <- check
+          case currentStatus of
+              Success -> print ("reloading: " <> desc) >> reload
+              _       -> print ("turning-up: " <> desc) >> turnup
+          print ("turnup-done: " <> desc)
+          bcast (oid,Success,Stable,TurnedUp)
 
 -- | Turn-down a graph sequentially
--- Topologically sort all nodes of the graph then run corresponding operations
+--
+-- Topologically sorts a transposition of the graph then invokes `TurnDown` action synchronously
+-- through each node. Note the order of execution is deterministic.
 syncTurnDownGraph :: Broadcast -> OpGraph -> IO ()
 syncTurnDownGraph bcast (graph,lookupVertex,_) =
-  mapM_ go (reverse $ Graph.topSort graph)
-  where
-    go :: Graph.Vertex -> IO ()
-    go vertex = do
-        let (preop,oid,_) = lookupVertex vertex
-            desc          = opName . opDescription $ runPreOp preop
-            turndown      = opTurndown $ opFunctions $ runPreOp preop
-        bcast (oid,Unknown,Transient,TurnedDown)
-        print ("turning-down: "<> desc)
-        turndown
-        print ("turndown-done: " <> desc)
-        bcast (oid,Success,Stable,TurnedDown)
+    mapM_ go (Graph.topSort $ Graph.transposeG graph)
```

Comment:
I don't remember which is which but you need to transposeG only at TurnUp or at TurnDown, your presently doing both with `mapM_ go (Graph.topSort $ Graph.transposeG graph)` . It's not the end of the world but I'm under the impression that this would remove the table before removing the cutlery and plates.